codec/codec.go

```diff
@@ -0,0 +1,46 @@
+package codec
```

Comment:
not sure whether you got chance reading though my reply before sending this. The real concern is where to cache the codec to simplify the logic.

To your question, yes but it is not the reason I took the current approach.